File: BraintreeDataCollector/src/androidTest/java/com/braintreepayments/api/DataCollectorTest.java
package com.braintreepayments.api;

import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.fail;

import android.content.Context;

import androidx.annotation.Nullable;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.concurrent.CountDownLatch;

@RunWith(AndroidJUnit4ClassRunner.class)
public class DataCollectorTest {

    private Context context;

    private CountDownLatch countDownLatch;

    @Before
    public void beforeEach() {
        context = ApplicationProvider.getApplicationContext();
        countDownLatch = new CountDownLatch(1);
    }

    @After
    public void afterEach() {
        SharedPreferencesHelper.clearConfigurationCacheOverride(context);
    }

    @Test(timeout = 10000)
    public void collectDeviceData_returnsCorrelationIdAndDoesNotCollectKountData() throws InterruptedException, InvalidArgumentException {
        Configuration configuration = new TestConfigurationBuilder().buildConfiguration();
        Authorization authorization = Authorization.fromString(Fixtures.TOKENIZATION_KEY);

        SharedPreferencesHelper.overrideConfigurationCache(context, authorization, configuration);

        BraintreeClient braintreeClient = new BraintreeClient(context, Fixtures.TOKENIZATION_KEY);
        DataCollector sut = new DataCollector(braintreeClient);

        sut.collectDeviceData(context, new DataCollectorCallback() {
            @Override
            public void onResult(@Nullable String deviceData, @Nullable Exception error) {
                try {
                    assertNotNull(deviceData);
                    JSONObject json = new JSONObject(deviceData);
                    assertNull(Json.optString(json, "device_session_id", null));
                    assertNull(Json.optString(json, "fraud_merchant_id", null));
                    assertNotNull(json.getString("correlation_id"));
                    countDownLatch.countDown();
                } catch (JSONException e) {
                    fail(e.getMessage());
                }
            }
        });

        countDownLatch.await();
    }
}

File: DataCollector/src/androidTest/java/com/braintreepayments/api/datacollector/DataCollectorTest.kt
package com.braintreepayments.api.datacollector

import android.text.TextUtils
import androidx.test.core.app.ApplicationProvider
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import com.braintreepayments.api.core.Authorization
import com.braintreepayments.api.core.Configuration
import com.braintreepayments.api.testutils.Fixtures
import com.braintreepayments.api.testutils.SharedPreferencesHelper
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.json.JSONObject
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.util.concurrent.CountDownLatch

@RunWith(AndroidJUnit4ClassRunner::class)
class DataCollectorTest {

    private lateinit var configuration: Configuration

    @Before
    fun setUp() {
        configuration = Configuration.fromJson(Fixtures.CONFIGURATION_WITH_LIVE_PAYPAL)
        val authorization = Authorization.fromString(Fixtures.TOKENIZATION_KEY)
        SharedPreferencesHelper.overrideConfigurationCache(
            ApplicationProvider.getApplicationContext(),
            authorization,
            configuration
        )
    }

    @Test
    fun getClientMetadataId_returnsClientMetadataId() {
        val sut = DataCollector(
            ApplicationProvider.getApplicationContext(),
            Fixtures.TOKENIZATION_KEY
        )
        val clientMetadataId = sut.getClientMetadataId(
            ApplicationProvider.getApplicationContext(),
            configuration,
            true
        )

        assertFalse(TextUtils.isEmpty(clientMetadataId))
    }

    @Test(timeout = 10000)
    fun collectDeviceData_returnsSuccessWithCorrelationId() {
        val countDownLatch = CountDownLatch(1)
        val sut = DataCollector(
            ApplicationProvider.getApplicationContext(),
            Fixtures.TOKENIZATION_KEY
        )
        val request = DataCollectorRequest(hasUserLocationConsent = true)

        sut.collectDeviceData(
            ApplicationProvider.getApplicationContext(),
            request
        ) { result ->
            assertTrue(result is DataCollectorResult.Success)
            val deviceData = (result as DataCollectorResult.Success).deviceData
            val json = JSONObject(deviceData)
            assertFalse(TextUtils.isEmpty(json.getString("correlation_id")))
            countDownLatch.countDown()
        }

        countDownLatch.await()
    }

    @Test(timeout = 10000)
    fun collectDeviceData_withRiskCorrelationId_returnsSuccess() {
        val countDownLatch = CountDownLatch(1)
        val sut = DataCollector(
            ApplicationProvider.getApplicationContext(),
            Fixtures.TOKENIZATION_KEY
        )
        val request = DataCollectorRequest(
            hasUserLocationConsent = true,
            riskCorrelationId = "custom-correlation-id"
        )

        sut.collectDeviceData(
            ApplicationProvider.getApplicationContext(),
            request
        ) { result ->
            assertTrue(result is DataCollectorResult.Success)
            val deviceData = (result as DataCollectorResult.Success).deviceData
            val json = JSONObject(deviceData)
            assertFalse(TextUtils.isEmpty(json.getString("correlation_id")))
            countDownLatch.countDown()
        }

        countDownLatch.await()
    }

    @Test(timeout = 10000)
    fun collectDeviceData_whenConfigurationFetchFails_returnsFailure() {
        val countDownLatch = CountDownLatch(1)
        val sut = DataCollector(
            ApplicationProvider.getApplicationContext(),
            "sandbox_invalid_key_xxxxxxxx"
        )
        val request = DataCollectorRequest(hasUserLocationConsent = true)

        sut.collectDeviceData(
            ApplicationProvider.getApplicationContext(),
            request
        ) { result ->
            assertTrue(result is DataCollectorResult.Failure)
            countDownLatch.countDown()
        }

        countDownLatch.await()
    }

    @Test(timeout = 10000)
    fun collectDeviceDataOnSuccess_returnsSuccessWithCorrelationId() {
        val countDownLatch = CountDownLatch(1)
        val sut = DataCollector(
            ApplicationProvider.getApplicationContext(),
            Fixtures.TOKENIZATION_KEY
        )
        val request = DataCollectorRequest(hasUserLocationConsent = true)

        sut.collectDeviceDataOnSuccess(
            ApplicationProvider.getApplicationContext(),
            request
        ) { result ->
            assertTrue(result is DataCollectorResult.Success)
            val deviceData = (result as DataCollectorResult.Success).deviceData
            val json = JSONObject(deviceData)
            assertFalse(TextUtils.isEmpty(json.getString("correlation_id")))
            countDownLatch.countDown()
        }

        countDownLatch.await()
    }

    @Test(timeout = 10000)
    fun collectDeviceDataOnSuccess_withRiskCorrelationId_returnsSuccess() {
        val countDownLatch = CountDownLatch(1)
        val sut = DataCollector(
            ApplicationProvider.getApplicationContext(),
            Fixtures.TOKENIZATION_KEY
        )
        val request = DataCollectorRequest(
            hasUserLocationConsent = true,
            riskCorrelationId = "custom-correlation-id"
        )

        sut.collectDeviceDataOnSuccess(
            ApplicationProvider.getApplicationContext(),
            request
        ) { result ->
            assertTrue(result is DataCollectorResult.Success)
            val deviceData = (result as DataCollectorResult.Success).deviceData
            val json = JSONObject(deviceData)
            assertFalse(TextUtils.isEmpty(json.getString("correlation_id")))
            countDownLatch.countDown()
        }

        countDownLatch.await()
    }

    @Test(timeout = 10000)
    fun collectDeviceDataOnSuccess_whenConfigurationFetchFails_returnsFailure() {
        val countDownLatch = CountDownLatch(1)
        val sut = DataCollector(
            ApplicationProvider.getApplicationContext(),
            "sandbox_invalid_key_xxxxxxxx"
        )
        val request = DataCollectorRequest(hasUserLocationConsent = true)

        sut.collectDeviceDataOnSuccess(
            ApplicationProvider.getApplicationContext(),
            request
        ) { result ->
            assertTrue(result is DataCollectorResult.Failure)
            countDownLatch.countDown()
        }

        countDownLatch.await()
    }
}
